js/app/test/blocks_chained_events.spec.ts

```diff
@@ -0,0 +1,61 @@
+import { test, expect } from "@gradio/tootils";
+
+test("Success event", async ({ page }) => {
+	const textbox = page.getByLabel("Result");
+	await expect(textbox).toHaveValue("");
+
+	await Promise.all([
+		page.click("text=Trigger Failure"),
+		page.waitForResponse("**/run/predict")
+	]);
+	// Since the event is not triggered, the value should not change
+	expect(textbox).toHaveValue("");
+
+	await Promise.all([
+		page.click("text=Trigger Success"),
+		page.waitForResponse("**/run/predict")
+	]);
+
+	expect(textbox).toHaveValue("Success event triggered");
+});
+
+test("Consecutive success event", async ({ page }) => {
+	const textbox = page.getByLabel("Consecutive Event");
+
+	await page.click("text=Trigger Consecutive Success");
+	// need to wait for all the events to fire
+	await page.waitForTimeout(1000);
+
+	expect(textbox).toHaveValue("Consecutive Event Triggered");
+});
+
+test("gr.Error Triggers modal", async ({ page }) => {
+	await Promise.all([
+		page.click("text=Trigger Failure"),
+		page.waitForResponse("**/run/predict")
+	]);
+
+	const toast = page.getByTestId("error-toast");
+	expect(toast).toContainText("Something went wrong");
+	const close = page.getByTestId("error-close");
+	await close.click();
+	// need to let the toast fade out
+	await page.waitForTimeout(2000);
+	expect(await page.getByTestId("error-toast").count()).toEqual(0);
```

Comment:
Instead of a timeout which could fail if the animation takes a little longer, we can wait for the element to vanish and await the assertion:

```suggestion
	await expect(page.getByTestId("error-toast")).toHaveCount(0);

```

Reply:
Great suggestion! But confused how this works. How does `await expect(page.getByTestId("error-toast")).toHaveCount(0);` mean that we'll wait for the element to disappear?

Reply:
Playwright has this philosophy of trying to avoid flake at all costs. The reason the 'locators' return promises (and need to be awaited) is because they will keep trying to find an element. 

In this case instead of awaiting the locator we pass the 'attempt at location' into expect. When you do that you await the expectation will will retry until it is true. 

These attempts at location will not hang indefinitely, they use the global timeout (30sec by default I think) after which the test will fail. You can set a custom timeout by doing `toHaveCount(0, { timeout: 5000 })`.

Reply:
Very nice. Thank you for explaining!